crates/core/src/client/client_api.rs

```diff
@@ -1,126 +1,97 @@
+use async_trait::async_trait;
 use eyre::Result;
 // TODO: all reth_primitives::rpc types should be replaced when native reth Log is implemented
 // https://github.com/paradigmxyz/reth/issues/1396#issuecomment-1440890689
 use reth_primitives::{Address, BlockId, BlockNumberOrTag, Bytes, H256, U128, U256, U64};
 use reth_rpc_types::{
-    BlockTransactions, CallRequest, FeeHistory, RichBlock, SyncStatus, Transaction as EtherTransaction,
+    BlockTransactions, CallRequest, FeeHistory, Index, RichBlock, SyncStatus, Transaction as EtherTransaction,
     TransactionReceipt,
 };
 use starknet::core::types::{BlockId as StarknetBlockId, BroadcastedInvokeTransactionV1, FieldElement};
-use starknet::providers::jsonrpc::{HttpTransport, JsonRpcClientError};
-use starknet::providers::{JsonRpcClient, ProviderError};
-use thiserror::Error;
-extern crate hex;
+use starknet::providers::jsonrpc::HttpTransport;
+use starknet::providers::JsonRpcClient;
 
-use async_trait::async_trait;
-use reth_rpc_types::Index;
-
-use super::helpers::DataDecodingError;
-use crate::models::{ConversionError, StarknetTransactions, TokenBalances};
-
-#[derive(Debug, Error)]
-pub enum KakarotClientError {
-    #[error(transparent)]
-    RequestError(#[from] ProviderError<JsonRpcClientError<reqwest::Error>>),
-    #[error(transparent)]
-    ConversionError(#[from] ConversionError),
-    #[error(transparent)]
-    DataDecodingError(#[from] DataDecodingError),
-    #[error(transparent)]
-    OtherError(#[from] anyhow::Error),
-}
+use super::errors::EthApiError;
+use crate::models::{StarknetTransactions, TokenBalances};
 
 #[async_trait]
 pub trait KakarotClient: Send + Sync {
     fn kakarot_address(&self) -> FieldElement;
     fn proxy_account_class_hash(&self) -> FieldElement;
     fn inner(&self) -> &JsonRpcClient<HttpTransport>;
 
-    async fn block_number(&self) -> Result<U64, KakarotClientError>;
+    async fn block_number(&self) -> Result<U64, EthApiError>;
 
-    async fn transaction_by_hash(&self, hash: H256) -> Result<EtherTransaction, KakarotClientError>;
+    async fn transaction_by_hash(&self, hash: H256) -> Result<EtherTransaction, EthApiError>;
 
     async fn get_eth_block_from_starknet_block(
         &self,
         block_id: StarknetBlockId,
         hydrated_tx: bool,
-    ) -> Result<RichBlock, KakarotClientError>;
+    ) -> Result<RichBlock, EthApiError>;
 
     async fn get_code(
         &self,
         ethereum_address: Address,
         starknet_block_id: StarknetBlockId,
-    ) -> Result<Bytes, KakarotClientError>;
+    ) -> Result<Bytes, EthApiError>;
 
     async fn call_view(
         &self,
         ethereum_address: Address,
         calldata: Bytes,
         starknet_block_id: StarknetBlockId,
-    ) -> Result<Bytes, KakarotClientError>;
+    ) -> Result<Bytes, EthApiError>;
```

Comment:
I feel like `EthApiError` may be a confusing name. Here for instance, we call Starknet with a "call" request, it feels weird that we may return an `EthApiError`.

Could we name it `KakarotError`? Or something else, able to tell people this error comes from the core business logic of Kakarot.

Reply:
Alternatively, we could split our core business logic into:
StarknetWrapper trait and
KakarotEthApi trait, and then have KakarotClient be a composition of both traits,

Then, you can have EthApiError for KakarotEthApi trait, that only implements functions that have ~almost the same signature as Ethereum JSON RPC client calls.

Then, we can have StarknetWrapperError for StarknetWrapper trait, that only implements function that wrap starknet calls but format things to adapt to Kakarot business logic.